Add confirmation before certificate delete

osirisinferi opened the pull request:
Fixes #8347, #8348

Not sure about:

* The indenting and marking of the listing of the to be removed certificates. Two spaces with an asterisk seemed logical to *me* personally, but other opinions or established routines might prevail.
* The added tests keep the `cert_manager.py` coverage at the same level, but please, take a good look if I at least did a decent job with those things.. I am *not* a fan of writing, errr, hacking and copy/pasting together those things.

alexzorin commented on certbot/certbot/_internal/cert_manager.py:
```diff
@@ -90,9 +90,23 @@ def certificates(config):
+    for certname in certnames:
+        logger.info("  * " + certname)
+    if not disp.yesno("Are you sure to delete the above certificate{0}?".
+        format(suffix), force_interactive=True):
```
The ability to non-interactively delete certificates needs to be preserved. For example, when Certbot is being scripted, this errors unconditionally:

    $ certbot -n delete --cert-name localhost
    Saving debug log to /var/log/letsencrypt/letsencrypt.log

    The following certificate is selected for deletion:

      * localhost
    Missing command line flag or config entry for this setting:
    Are you sure to delete the above certificate?

Maybe `force_interactive=False, default="Yes"`?

osirisinferi replied:
Hmm, I thought I looked at that. I definitely looked at `cli_flag` where I also seemed to remember the existance of some kind of `--yes-to-all` flag.. But I could also be mistaken with another piece of software :stuck_out_tongue: 

I'll try `default="Yes"`.. It might perhaps also fix the integration test failure.. `tox -e py3-cover` ran perfectly locally though.

alexzorin reviewed (changes requested):
I agree with this change because the destruction of certificates is not necessarily something a user can easily recover from.

The "delete all my certificates" case is a no-brainer for a prompt.

It is a tad redundant to prompt for confirmation even when a singular `--cert-name` is specified. But pondering all the times `rm` being default aliased to `rm -i` has saved me, perhaps erring on the side of caution is best.

alexzorin commented on certbot/certbot/_internal/cert_manager.py:
```diff
@@ -90,9 +90,23 @@ def certificates(config):
+    else:
+        suffix = ""
+        verb = "is"
+    logger.info("\nThe following certificate%s %s selected for deletion:\n",
+        suffix, verb)
+    for certname in certnames:
+        logger.info("  * %s", certname)
```
Could you please join all of this output into the `yesno` prompt, instead of using `logger`?  

[Here](https://github.com/certbot/certbot/blob/ac8798e8185699cd2b7ddc904bccd978968e8e2f/certbot/certbot/_internal/auth_handler.py#L444-L469) is an example.

osirisinferi commented on certbot/certbot/_internal/cert_manager.py:
```diff
@@ -90,9 +90,23 @@ def certificates(config):
     """Delete Certbot files associated with a certificate lineage."""
     certnames = get_certnames(config, "delete", allow_multiple=True)
+    disp = zope.component.getUtility(interfaces.IDisplay)
+    if len(certnames) > 1:
+        suffix = "s"
+        verb = "are"
+    else:
+        suffix = ""
+        verb = "is"
```
```
(venv3) osiris@desktop certbot-test-environment $ grep -RIi "certificate(s)" certbot*
certbot/docs/using.rst:the certificate(s) to be ignored when considering renewal, and attempts to
certbot/tests/cert_manager_test.py:        prompt = "Which certificate(s) would you"
certbot/certbot/_internal/cert_manager.py:                prompt = "Which certificate(s) would you like to {0}?".format(verb)
(venv3) osiris@desktop certbot-test-environment $ 
```
Not much trouble to modify just `get_certnames` too I guess? I wouldn't mind using `certificate(s)` if it weren't for the `is/are" bit.

Also, your suggestion is obviously better, but higher Python for me at my current "level of Python knowledge" :grin:

osirisinferi commented:
@alexzorin Addressed your comments. I've left `logger.info()` for when the user choose "No". Perhaps it might be better to chose `warn()` though, not sure.

Also, I've chosen to expand the `certificate(s)` in `cert_names()` to keep everything in line. If the general opinion rules against this, I'll revert this. If that's the case, I'd like a better phrasing than "(&hellip;) certificate(s) is/are (&hellip;), so in that case better suggestions are welcome.

alexzorin commented on certbot/certbot/_internal/cert_manager.py:
```diff
@@ -90,9 +90,19 @@ def certificates(config):
+        .format(suffix, verb)]
+    for certname in certnames:
+        msg.append("  * " + certname)
+    msg.append("\nAre you sure to delete the above certificate{0}?".
```
Maybe "are you sure you want to"?

alexzorin replied:
Perhaps:

```suggestion
    msg.append("\nAre you sure you want to delete the above certificate{0}?".
```

alexzorin commented on certbot/certbot/_internal/cert_manager.py:
```diff
@@ -305,7 +315,8 @@ def get_certnames(config, verb, allow_multiple=False, custom_prompt=None):
         if allow_multiple:
             if not custom_prompt:
-                prompt = "Which certificate(s) would you like to {0}?".format(verb)
+                prompt = ("Which certificate or certificates would you like to "
```
I don't see a reason to expand this phrasing, personally.  If it is a dislike of `(s)`, I think prompting "Which certificates would you like to delete?" make sense, even if there is only a single certificate available or chosen.

alexzorin reviewed (changes requested):
LGTM otherwise

alexzorin commented on certbot/certbot/_internal/cert_manager.py:
```diff
@@ -90,9 +90,16 @@ def certificates(config):
     """Delete Certbot files associated with a certificate lineage."""
     certnames = get_certnames(config, "delete", allow_multiple=True)
+    disp = zope.component.getUtility(interfaces.IDisplay)
+    msg = ["The following certificate(s) is/are selected for deletion:\n"]
```
Sorry for the delay! Let's get this merged.

I think it's OK to just use the plural form here.

```suggestion
    msg = ["The following certificate(s) are selected for deletion:\n"]
```
